### PR TITLE
Rlecomte/tdi 44222 ftp connection pool

### PR DESCRIPTION
**Pull request by rlecomte-talend**
https://jira.talendforge.org/browse/TDI-44222
Added a connection pool in FTPService to reduce the number of created connections.
Some Unit test are disabled because of an issue with permissions checking. This will be resolved in another PR.

### CODE REVIEW

File: ftp/src/main/java/org/talend/components/ftp/service/FTPService.java [hunk 1]
```diff
@@ -99,7 +107,22 @@ public String generateFilePath(String dirPath, String filename) {
         return dirPath + (dirPath.endsWith(PATH_SEPARATOR) ? "" : PATH_SEPARATOR) + filename;
     }
 
-    public GenericFTPClient getClient(FTPDataStore dataStore) {
+    public GenericFTPClient getClient(FTPConnectorConfiguration configuration) {
+
+        if (clientMap == null) {
+            clientMap = Collections.synchronizedMap(new HashMap<>());
```
Comment (clesaec) [inline ftp/src/main/java/org/talend/components/ftp/service/FTPService.java:113]: Use class java.util.concurrent.ConcurrentHashMap
=> better optimized for conccurence.

File: ftp/src/main/java/org/talend/components/ftp/service/FTPService.java [hunk 2]
```diff
@@ -99,7 +107,22 @@ public String generateFilePath(String dirPath, String filename) {
...
+        }
+
+        GenericFTPClient client = clientMap.get(configuration.getConfigKey());
+        if (client == null || !client.isConnected()) {
+            client = getClient(configuration.getDataSet().getDatastore());
+            clientMap.put(configuration.getConfigKey(), client);
+        }
```
Comment (clesaec) [inline ftp/src/main/java/org/talend/components/ftp/service/FTPService.java:120]: Use Map.compute(K key,
            BiFunction<? super K, ? super V, ? extends V> remappingFunction) here.
remapping function could simply be a private function :
```
private GenericFTPClient computeClient(String key, GenericFTPClient current) {
       if (current == null || !current.isConnected()) {
            return  getClient(configuration.getDataSet().getDatastore());
       }
       return current
}
```
and call `GenericFTPClient client = clientMap.compute(configuration.getConfigKey(), this::computeClient);`
So, it respect multi-threading contract

File: ftp/src/test/java/org/talend/components/ftp/output/FTPOutputTest.java [hunk 3]
```diff
@@ -93,6 +96,39 @@ void buildConfig() {
         configuration.setDebug(true);
     }
 
+    // @EnvironmentalTest
```
Comment (clesaec) [inline ftp/src/test/java/org/talend/components/ftp/output/FTPOutputTest.java:99]: Can't be active ?

File: ftp/src/main/java/org/talend/components/ftp/service/FTPService.java (hunk 2 above)
Comment (rlecomte-talend) [reply ftp/src/main/java/org/talend/components/ftp/service/FTPService.java:120]: I did that on the first occurence, but I had to add the isConnected test.

Comment (clesaec) [reply ftp/src/main/java/org/talend/components/ftp/service/FTPService.java:120]: My computeClient example function also has "isConnected" test

Comment (rlecomte-talend) [reply ftp/src/main/java/org/talend/components/ftp/service/FTPService.java]: OK, done

**Pull request merged**